### PR TITLE
Cleaning up error types

### PR DESCRIPTION
**Pull request by robertbastian**
This PR removes some error types that are only around so that their `From` implementation can be used implicitely once or twice. This is especially confusing if types wrap `DataError`s and also implement `Into<DataError>`, as at the call site it becomes hard to figure out how the error is actually being transformed.

I also removed `buffer_type` from `DataResponseMetadata` and made `BufferProvider::load_buffer` return it. This information is only relevant to `BufferProvider`, and should always be returned from it (missing it was a runtime error before).

Edit: also removed `DataError::with_error_context` because it's confusing. As `std::error::Error` is not `Copy`, the most we can do with an error is log it. But we already have `with_display_context` for this, which does the exact same thing and works for all errors (`std::error::Error : Display`) regardless of whether `std` is available.

### CODE REVIEW

**Review by Manishearth — COMMENTED** [review]
I'm not quite sure I understand what's going on here, and some of this seems to be undoing explicit choices by @sffc in past PRs, I'll defer to him here

File: provider/core/src/error.rs
Comment (robertbastian) [reply provider/core/src/error.rs:283]: `crate::serde::Error` was one of these intermediate error types which were immediately converted to `DataError`. `postcard:error` had two paths, `postcard::error` -> `crate::serde::Error` -> `DataError`, and `postcard::Error` -> `DataError`. Removing the `From` impls makes it explicit what is going on.

File: provider/core/src/serde/de.rs
Comment (sffc) [inline provider/core/src/serde/de.rs:50]: Issue: Please restore this as a regular function. I wrote it as a regular function because

1. It is a complex function and I strongly prefer for complex functions to be their own function
2. It reduces indentation which allows for more content on each line
3. This is the type of function that may often show up in stack traces and it's helpful if it has its own name
4. The fact that the function has a type argument is more clear
5. If we needed to share this function, we could do so as a standalone function
6. During DCE, I'd be happy if all of the DataProvider code got inlined, but I think it may be useful to keep this as the "terminal" function that doesn't get inlined, especially if we need to deserialize the same type `M` in multiple places in the same program

File: provider/core/src/serde/mod.rs
Comment (sffc) [inline provider/core/src/serde/mod.rs:44]: Praise: I support getting rid of this extra error type

File: provider/core/src/error.rs
Comment (sffc) [inline provider/core/src/error.rs:240]: Praise: It's fine to get rid of this one since, as you said, we're only using the `Display` impl from the error.

File: provider/fs/src/export/serializers/mod.rs
Comment (sffc) [inline provider/fs/src/export/serializers/mod.rs:16]: Thought: IIRC, one motivation of having separate internal error types like this one was to prevent accidentally forgetting to associate the path with the error, since in order to convert from `serializers::Error` to `DataError` you needed to attach a path. But it's a lot of boilerplate to achieve that vague objective, so I am supportive of this simplication.

File: provider/datagen/src/error.rs
Comment (sffc) [inline provider/datagen/src/error.rs:12]: Praise: Although I had originally anticipated moving all of `icu_datagen` to `eyre`, I think it makes sense today to move it all to `DataError` instead, since so much work happens inside the DataProvider impls, which return `DataError`.

**Review by sffc — CHANGES_REQUESTED** [review]
Praise: I support collapsing the various different error types into DataError and leveraging it everywhere, which is the main purpose of this PR.

I'm not as supportive of the tangential changes that made it into the PR such as returning BufferFormat in a tuple.

File: provider/core/src/data_provider.rs
Comment (sffc) [reply provider/core/src/data_provider.rs:47]: I'm willing to discuss the proposed change to BufferProvider return values, but please decouple it from this PR which is about error types.

> Without `BufferFormat`, the `[u8]` is useless

It's not hard to imagine systems that infer the format or use some mechanism other than serde for deserializing the buffer.

**Review by sffc — CHANGES_REQUESTED** [review]
Please decouple the BufferProvider change from the rest of the PR

**Comment by sffc** [comment]
Also, I think this PR is likely to cause a merge conflict with the collator PR. Let's hold off on merging this until collator is merged.

**Comment by robertbastian** [comment]
> Please decouple the BufferProvider change from the rest of the PR

Done

I've also made some tweaks to `SerializeBox`. `BlobDataExporter` was using its `serialize` API, while `FsDataExporter` was using its `Deref<dyn erased_serde::Serialize>` impl, so the error handling between the two was different. I've changed `FsDataExporter` over to the `serialize` API and updated it so that the callers don't have to use `erased_serde`.

File: provider/core/src/serde/ser.rs
Comment (sffc) [inline provider/core/src/serde/ser.rs:56]: Praise: Good idea bringing the erased_serde interface down one more layer, so that child crates don't need the dependency

File: provider/fs/src/export/serializers/mod.rs
Comment (sffc) [inline provider/fs/src/export/serializers/mod.rs:19]: Nit (optional): since you don't need ownership, take `&DataPayload<SerializeMarker>` or `&SerializeBox` (the return value of `DataPayload::get()`)

**Review by sffc — DISMISSED** [review]

File: provider/fs/src/export/serializers/mod.rs
Comment (robertbastian) [reply provider/fs/src/export/serializers/mod.rs:19]: I'll leave it like this for now but I have a further change here in the queue.

**Review by sffc — APPROVED** [review]
Still LGTM

**Pull request merged**
